src/lib/ares_private.h

```diff
@@ -315,6 +323,20 @@ struct ares_channeldata {
 
   /* Query Cache */
   ares__qcache_t                     *qcache;
+
+  /* Callback triggered when a server has a successful or failed response. */
+  ares_server_state_callback server_state_cb;
+  void                      *server_state_cb_data;
+
+  /* Fields controlling server failover behavior.
+   * The retry chance is the probability (1/N) by which we will retry a failed
```

Comment:
pretty sure the probability isn't 1/N, as values 128-255 have the same probability of hitting, 2 in 256 (0 and itself). 86-127 have the same probability as well, 3 in 256. 64-85 have the same probability, 4 in 256.

Seems logarithmic, but haven't thought hard enough to say what the algorithm really is.

Reply:
Ah, good spot! The probability will be ceil(256/N)/256, which is 1/N rounded up to the nearest 1/256. I think this is best we can do with a single random byte, as we only have 8 bits of precision.

Perhaps we can use 4 random bytes, which would give us probabilities to the nearest 1/2^32. That feels much better in terms of precision and we could always document the specific behaviour.

Reply:
I have moved the retry chance to use `size_t`s, which should give increased precision. I think the precision will be small enough that external documentation is not hugely important, although happy to add a note when the man pages are written.

Reply:
I'm maybe slightly worried about draining the RNG since it does require a syscall to fetch more bytes and it is cryptographically secure which means its much slower than say something like xorshift (which would be sufficient for something like this).  At least we do pull from the OS in 256byte chunks these days.

Maybe grabbing 2 bytes (16bits) would be a good compromise?

Reply:
Ok thanks, I think 1/2^16 will be more than enough precision. I have switch this to use an `unsigned short`.